File: tutorial/tutsim.py
# Python version of tutsim.cc
# Implements a very basic agent based model
# Code can be used as basis for more complex models.

import numpy
import random

YEAR = 365.25
MALE = 0
FEMALE = 1

# Get the same random numbers every time we run. 23 is arbitrary
numpy.random.seed(23)
random.seed(23)


class Agent:

    def __init__(self):
        self.sex = random.randint(MALE, FEMALE)
        self.age = random.uniform(15.0, 20.0)
        self.hiv = min(numpy.random.geometric(p=0.9) - 1, 5)

    def infection_event(self, prevalence, prob_new_partner, force_infection):
        if self.hiv == 0:
            risk_infection = force_infection * prob_new_partner * prevalence
            if random.random() < risk_infection:
                self.hiv = 1

    def age_event(self, time_elapsed):
        self.age += time_elapsed


def report(date, agents):
    infected = len([a.hiv for a in agents if a.hiv > 0])
    prevalence = infected / len(agents)
    print("{0} Num infected: {1} Prevalence: {2}".
          format(date, infected, prevalence))


def simulate(agents, parameters):
    num_iterations = int(parameters['NUM_YEARS'] / parameters['TIME_STEP'])
    for i in range(num_iterations):
        random.shuffle(agents)
        prevalence = len([a.hiv for a in agents if a.hiv > 0]) / len(agents)
        for agent in agents:
            agent.infection_event(prevalence,
                                  parameters["PROB_NEW_PARTNER"],
                                  parameters["FORCE_INFECTION"])
            agent.age_event(parameters["TIME_STEP"])
        report(parameters["START_DATE"] + i / YEAR, agents)


def print_verbose_agent_info(agents):
    males = [a.sex for a in agents].count(MALE)
    avg_age = sum([a.age for a in agents]) / len(agents)
    youngest = min([a.age for a in agents])
    oldest = max([a.age for a in agents])
    hiv_counts = [[a.hiv for a in agents].count(i) for i in range(6)]
    print("DEBUG", hiv_counts)
    print("Males: ", males)
    print("Youngest: ", youngest)
    print("Oldest: ", oldest)
    print("Average age: ", avg_age)
    for i in range(6):
        print("HIV ", i, hiv_counts[i])


def main():
    parameters = {
        'NUM_YEARS': 2.0,
        'TIME_STEP': 1.0 / YEAR,
        'START_DATE': 2015.0,
        'PROB_NEW_PARTNER': 0.022,
        'FORCE_INFECTION': 0.1
    }
    agents = [Agent() for _ in range(100)]
    print_verbose_agent_info(agents)
    simulate(agents, parameters)
    print_verbose_agent_info(agents)

if __name__ == "__main__":
    main()

File: tutorial/tutsim.R
make_agents <- function(num_agents, min_age=15, max_age=20) {
  gender <- factor(c("male","female"))
  agents <- data.frame(id=c(1:num_agents), 
                       sex=sample(gender, num_agents, replace=T), 
                       age=runif(num_agents,min_age,max_age), 
                       hiv=rgeom(num_agents, 0.9))
  return(agents)
}

agent_events <- function(agents, parameters) {
  # I don't understand the purpose of the shuffle step - it shouldn't affect the individual SC probabilities
  # Increment age
  agents$age <- agents$age + parameters$time_step
  # Infections
  prevalence <- sum(agents$hiv > 0) / nrow(agents)
  risk_infection <- parameters$force_infection * parameters$prob_new_partner * prevalence
  agents$hiv[agents$hiv==0] <- rbinom(length(agents$hiv[agents$hiv==0]),1,risk_infection)
  return(agents)
}

simulate <- function(agents, parameters) {
  time <- seq(from=parameters$start_date,
              to=parameters$start_date+parameters$num_years,
              by=parameters$time_step)
  for (t in time) {
    agents <- agent_events(agents, parameters)
    num_hiv <- sum(agents$hiv > 0)
    prevalence <- num_hiv / nrow(agents)
    print(c(t,num_hiv,prevalence))
  }
  return(agents)
}

parameters <- list(start_date=2015,
                num_years=2,
                time_step=1/365,
                prob_new_partner=0.022,
                force_infection=0.1)

agents <- make_agents(10000, min_age = 15, max_age = 20)
print(summary(agents))
agents <- simulate(agents, parameters)
print(summary(agents))

# Test how much time is taken by printing out results
system.time(simulate(agents, parameters))
simulate <- function(agents, parameters) {
  time <- seq(from=parameters$start_date,
              to=parameters$start_date+parameters$num_years,
              by=parameters$time_step)
  for (t in time) { agents <- agent_events(agents, parameters) }
  return(agents)
}
system.time(simulate(agents, parameters))